src/service/resources-impl.js

```diff
@@ -1292,6 +1275,110 @@ export class Resources {
       }
     }
   }
+
+  /**
+   * Calls iterator on each sub-resource
+   * @param {function(!Resource, number)} iterator
+   */
+  eachResource(iterator) {
+    const resources = this.resources_;
+    for (let i = 0; i < resources.length; i++) {
+      iterator(resources[i], i);
+    }
+  }
+
+  setupVisibilityStateMachine_(vsm) {
+    const viewerVisDocVis = {
+      viewer: true,
+      document: true
+    };
+    const viewerVisDocHid = {
+      viewer: true,
+      document: false
+    };
+    const viewerHidDocVis = {
+      viewer: false,
+      document: true
+    };
+    const viewerHidDocHid = {
+      viewer: false,
+      document: false
+    };
+
+    const doPass = () => {
+      // If viewport size is 0, the manager will wait for the resize event.
+      const viewportSize = this.viewport_.getSize();
+      if (viewportSize.height > 0 && viewportSize.width > 0) {
+        if (this.hasMutateWork_()) {
+          this.mutateWork_();
+        }
+        this.discoverWork_();
+        let delay = this.work_();
+        if (this.hasMutateWork_()) {
+          // Overflow mutate work.
+          delay = Math.min(delay, MUTATE_DEFER_DELAY_);
+        }
+        if (this.visible_) {
+          log.fine(TAG_, 'next pass:', delay);
+          this.schedulePass(delay);
+          this.updateScrollHeight_();
+        } else {
+          log.fine(TAG_, 'document is not visible: no scheduling');
+        }
+      }
+    };
+    const pause = () => {
+      this.eachResource(r => r.pause());
+    };
+    const unlayout = () => {
+      this.eachResource(r => r.unlayout());
+    };
+    const pauseAndUnlayout = () => {
+      this.eachResource(r => {
+        r.pause();
+        r.unlayout();
+      });
+    };
+    const resume = () => {
+      this.eachResource(r => {
+        r.resume();
+        doPass();
+      });
+    };
+    const layout = () => {
+      this.eachResource(r => {
+        r.layout();
+        doPass();
+      });
+    };
+    const resumeAndLayout = () => {
+      this.eachResource(r => {
+        r.resume();
+        r.layout();
+        doPass();
+      });
+    };
+
+    vsm.addTransition(viewerVisDocVis, viewerVisDocVis, doPass);
```

Comment:
Overall, I like the idea of transitions. But I believe that we need to have a single logic state if we can derive it and I think we should try. For instance, the states could be "prerendering", "active", "paused", "unloaded", etc.

Another "big" thing we need to consider is that all of these callbacks should be callable by "owner" elements such as carousel. I think we are pretty much there, but some per-resource state maybe needed.